### PR TITLE
chore(ecs): Refactor TaskHealthCachingAgent networking check

### PR DESCRIPTION
**Pull request by allisaurus**
Previously, how to check task health forked based on looking for network bindings on each container. Whether the task contains network bindings was already being determined in a previous check, so this surfaces that value and forks based on that instead. 

Also adds test to ensure tasks with missing or null network bindings/interfaces are skipped.

### CODE REVIEW

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/TaskHealthCachingAgent.java [hunk 1]
```diff
@@ -129,34 +129,38 @@ public TaskHealthCachingAgent(
...
-              "Task '{}' is missing networking. Will not retrieve health.", task.getTaskArn());
+              "Provided task definition '{}' is null for service '{}'",
+              service.getTaskDefinition(),
+              service.getServiceName());
+          continue;
+        }
+
+        boolean lacksNetworkBindings = isTaskMissingNetworkBindings(task);
+        if (task.getContainers().isEmpty()
+            || (lacksNetworkBindings && isTaskMissingNetworkInterfaces(task))) {
```
Comment (pkandasamy91) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/TaskHealthCachingAgent.java:142]: nit: we should probably also move isTaskMissingNetworkInterfaces to a variable (seems weird that lacksNetworkBindings is a variable while lacksNetworkInterfaces is not or vice versa)

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/TaskHealthCachingAgent.java [hunk 2]
```diff
@@ -129,34 +129,38 @@ public TaskHealthCachingAgent(
             Keys.getTaskDefinitionKey(accountName, region, service.getTaskDefinition());
         TaskDefinition taskDefinition = taskDefinitionCacheClient.get(taskDefinitionCacheKey);
 
-        if (isContainerMissingNetworking(task)) {
+        if (taskDefinition == null) {
```
Comment (pkandasamy91) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/TaskHealthCachingAgent.java:132]: Out of curiosity, why do we skip getting network bingings if taskDefinition does not exist in the cache? AFAICT NetworkBindings doesn't rely on the taskDefinition being present. Couldn't this lead to issues where taskDefinition hasn't yet been cached, and we skip the networkBindings/Interfaces check?

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/TaskHealthCachingAgent.java (hunk 1 above)
Comment (pkandasamy91) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/TaskHealthCachingAgent.java:142]: Also, technically a task should not be able to have both networkingBindings and networkInterfaces (can we add a check that logs an error if this were ever to happen - just so we're made aware of any weird behaviour) 

`          if (!lacksNetworkBindings && !lacksNetworkInterfaces) {`

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/TaskHealthCachingAgent.java [hunk 3]
```diff
@@ -129,34 +129,38 @@ public TaskHealthCachingAgent(
...
-          }
+        // ideally, could determine health check method by looking at taskDef.networkMode,
+        // however this isn't reliably cached yet, so reusing network binding check.
+        if (!lacksNetworkBindings) {
+          taskHealth =
+              inferHealthNetworkBindedContainer(
+                  targetHealthCacheClient, task, containerInstance, serviceName, service);
         }
 
         if (taskHealth == null) {
```
Comment (pkandasamy91) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/TaskHealthCachingAgent.java:158]: We should probably update this to be an else if condition since you can't have both networking bindings and interfaces present 

`else if (!lacksNetworkInterfaces) {`

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/TaskHealthCachingAgent.java [hunk 4]
```diff
@@ -178,11 +182,6 @@ private TaskHealth inferHealthNetworkInterfacedContainer(
       Service loadBalancerService,
       TaskDefinition taskDefinition) {
 
-    if (taskDefinition == null) {
```
Comment (pkandasamy91) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/TaskHealthCachingAgent.java:181]: Might still want to keep this based on my previous comment.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/TaskHealthCachingAgent.java [hunk 5]
```diff
@@ -251,12 +249,7 @@ private TaskHealth inferHealthNetworkBindedContainer(
       Task task,
       ContainerInstance containerInstance,
       String serviceName,
-      Service loadBalancerService,
-      TaskDefinition taskDefinition) {
-    if (taskDefinition == null) {
-      log.debug("Provided task definition is null.");
-      return null;
-    }
```
Comment (pkandasamy91) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/TaskHealthCachingAgent.java:259]: Good catch! We shouldn't be passing taskDefinition if we're not actually using it.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/TaskHealthCachingAgent.java (hunk 1 above)
Comment (allisaurus) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/TaskHealthCachingAgent.java:142]: originally i had both pulled out as vars, but the result of `isTaskMisingNetworkingInterfaces` was never used again, so I decided to just call the func here. Can put it back if there's a benefit to variable-izing it (not sure that's a word...)

Comment (pkandasamy91) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/TaskHealthCachingAgent.java:142]: I think it makes sense if it's used here as well https://github.com/spinnaker/clouddriver/pull/4712#discussion_r445765662

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/TaskHealthCachingAgent.java (hunk 2 above)
Comment (allisaurus) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/TaskHealthCachingAgent.java:132]: hmm you know originally both methods were returning if `taskDefinition` was null, so it made sense to just do once upfront. but since it's actually NOT being used by `inferHealthNetworkBindedContainer` and we can't use the network mode to determine how to health check (which would require it sooner - see comment on line 150), you're right that we can just do that check once below. Will update!

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/TaskHealthCachingAgent.java (hunk 3 above)
Comment (allisaurus) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/TaskHealthCachingAgent.java:158]: I wanted to keep the existing logic (outside removing the loop) mostly intact here, but I agree it's beneficial to log if both are found (see above). Maybe I can make this & the logging change in a separate PR so they're isolated?

Comment (pkandasamy91) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/TaskHealthCachingAgent.java:158]: Sounds good to me!
